### DISCUSSION TITLE
Examples of caching parsed document for completions, actions, symbols, folding, etc

### DISCUSSION THREAD

**Discussion by ptrthomas** [microsoft/vscode-discussions#273] [Extension Development QnA] [answered]
Is there a pattern for parsing the currently edited document once and caching it for use by all the providers such as `vscode.languages.registerCodeLensProvider()`, `registerCompletionItemProvider()`, `registerCodeActionsProvider()`, `registerDocumentSymbolProvider()`, `registerFoldingRangeProvider()` etc.

So if I re-parsed the active TextDocument on `vscode.workspace.onDidChangeTextDocument()` I want all the above routines to use the cache and not the passed-in document. Does this make sense? Or is this premature or un-necessary optimization because of the way VS Code works.

Do point me to good examples of how to handle this in your plugin / existing code.

**Comment by Austindgk232** [comment]
https://github.com/microsoft/vscode-discussions/discussions/273#discussion-4603868

**Comment by isidorn** [comment]
I am not sure. @jrieken might have a suggestion.

**Comment by jrieken** [comment, marked as answer]
I don't know of anything concrete but I'd assume most language brains use the `TextDocument#version`-property because that reflects the current version. Tho, please don't do much heavy lifting and parsing in the extension host but spawn a separate worker/process for that. Otherwise you are "steeling" compute resources from other extensions. Check out this guide for more: https://code.visualstudio.com/api/language-extensions/language-server-extension-guide

**Reply by ptrthomas** [reply]
Thanks, this helps !
